third_party/pcsc-lite/naclport/server_clients_management/src/client_request_processor.cc

```diff
@@ -223,17 +213,63 @@ void PcscLiteClientRequestProcessor::AsyncProcessRequest(
     std::shared_ptr<PcscLiteClientRequestProcessor> request_processor,
     RemoteCallRequestPayload request,
     RequestReceiver::ResultCallback result_callback) {
-  // This is paired with the call to `LeaveConcurrencyCheckScope()` in
-  // `ProcessRequest()`. We enter the scope here in order to be able to catch
-  // concurrency violations even when they occur so fast that the thread created
-  // below didn't have chance to start and do any work.
-  request_processor->EnterConcurrencyCheckScope(request.function_name);
-
   std::thread(&PcscLiteClientRequestProcessor::ProcessRequest,
               request_processor, std::move(request), result_callback)
       .detach();
 }
 
+PcscLiteClientRequestProcessor::ScopedConcurrencyGuard::ScopedConcurrencyGuard(
+    const std::string& function_name,
+    SCARDCONTEXT s_card_context,
+    SCARDHANDLE s_card_handle,
+    PcscLiteClientRequestProcessor& owner)
+    : owner_(owner),
+      function_name_(function_name),
+      s_card_context_(s_card_context) {
+  s_card_context_ =
+      s_card_handle
+          ? owner_.s_card_handles_registry_.FindContextByHandle(s_card_handle)
+          : s_card_context;
+  if (!s_card_context_) {
+    return;
+  }
+
+  std::string concurrent_functions_dump;
+  {
+    const std::unique_lock<std::mutex> lock(owner_.mutex_);
+    std::multiset<std::string>& running_functions =
+        owner_.context_to_running_functions_[s_card_context_];
+    concurrent_functions_dump =
+        JoinStrings(running_functions, /*separator=*/", ");
+    running_functions.insert(function_name_);
+  }
+
+  if (!concurrent_functions_dump.empty()) {
```

Comment:
This behaves wrong if we ever have a function with an empty string as a name, and it is the only running function. We should either check for empty names, or just use a separate bool to ascertain there are no running functions.

Reply:
Done - Used an `optional<>` here. (We couldn't have empty names today, but I agree it was fragile.)